src/cascadia/TerminalSettingsModel/AIConfig.cpp

```diff
@@ -0,0 +1,166 @@
+
+winrt::event_token AIConfig::AzureOpenAISettingChanged(const winrt::Microsoft::Terminal::Settings::Model::AzureOpenAISettingChangedHandler& handler) { return _azureOpenAISettingChangedHandlers.add(handler); };
+void AIConfig::AzureOpenAISettingChanged(const winrt::event_token& token) { _azureOpenAISettingChangedHandlers.remove(token); };
+
+winrt::hstring AIConfig::AzureOpenAIEndpoint() noexcept
+{
+    return _RetrieveCredential(PasswordVaultAIEndpoint);
+}
+
+void AIConfig::AzureOpenAIEndpoint(const winrt::hstring& endpoint) noexcept
+{
+    _SetCredential(PasswordVaultAIEndpoint, endpoint);
+    _azureOpenAISettingChangedHandlers();
+}
+
+winrt::hstring AIConfig::AzureOpenAIKey() noexcept
+{
+    return _RetrieveCredential(PasswordVaultAIKey);
+}
+
+void AIConfig::AzureOpenAIKey(const winrt::hstring& key) noexcept
+{
+    _SetCredential(PasswordVaultAIKey, key);
+    _azureOpenAISettingChangedHandlers();
+}
+
+static winrt::event<winrt::Microsoft::Terminal::Settings::Model::OpenAISettingChangedHandler> _openAISettingChangedHandlers;
+
+winrt::event_token AIConfig::OpenAISettingChanged(const winrt::Microsoft::Terminal::Settings::Model::OpenAISettingChangedHandler& handler) { return _openAISettingChangedHandlers.add(handler); };
+void AIConfig::OpenAISettingChanged(const winrt::event_token& token) { _openAISettingChangedHandlers.remove(token); };
+
+winrt::hstring AIConfig::OpenAIKey() noexcept
+{
+    return _RetrieveCredential(PasswordVaultOpenAIKey);
+}
+
+void AIConfig::OpenAIKey(const winrt::hstring& key) noexcept
+{
+    _SetCredential(PasswordVaultOpenAIKey, key);
+    _openAISettingChangedHandlers();
+}
+
+winrt::Microsoft::Terminal::Settings::Model::LLMProvider AIConfig::ActiveProvider()
+{
+    const auto val{ _getActiveProviderImpl() };
+    if (val)
+    {
+        // an active provider was explicitly set, return that
+        return *val;
+    }
+    else if (!AzureOpenAIEndpoint().empty() && !AzureOpenAIKey().empty())
+    {
+        // no explicitly set provider but we have an azure open ai key and endpoint, use that
+        return LLMProvider::AzureOpenAI;
+    }
+    else if (!OpenAIKey().empty())
+    {
+        // no explicitly set provider but we have an open ai key, use that
+        return LLMProvider::OpenAI;
+    }
+    else
+    {
+        return LLMProvider{};
+    }
+}
+
+void AIConfig::ActiveProvider(const LLMProvider& provider)
+{
+    _ActiveProvider = provider;
+}
+
+winrt::hstring AIConfig::_RetrieveCredential(const std::wstring_view credential)
+{
+    PasswordVault vault;
+    PasswordCredential cred;
+    // Retrieve throws an exception if there are no credentials stored under the given resource so we wrap it in a try-catch block
+    try
+    {
+        cred = vault.Retrieve(PasswordVaultResourceName, credential);
+    }
+    catch (...)
+    {
+        return L"";
+    }
+    return cred.Password();

```

Comment:
This seems expensive. Why not cache it on first load? Should the error be logged to aid in debugging?

Reply:
Added caching logic! Don't think we need to log it because it is expected behaviour that an exception gets thrown when the credential doesn't exist in the vault